[JSC] More sophisticated wasm worklist distribution

Constellation opened the pull request:
#### 2725a2a8434afdc0c01107f0396e7b9e7c948abc
<pre>
[JSC] More sophisticated wasm worklist distribution
<a href="https://bugs.webkit.org/show_bug.cgi?id=286986">https://bugs.webkit.org/show_bug.cgi?id=286986</a>
<a href="https://rdar.apple.com/144136820">rdar://144136820</a>

Reviewed by Keith Miller and Yijia Huang.

This patch tweaks our wasm worklist distribution code for non streaming
wasm compilation.

1. We check each function size and isolate super huge function in a
   different loop of wasm compilation so that we can distribute outlier
   sized function separately from the rest, which makes distribution of
   compilation task more uniformly.
2. We compute total size of functions and change partial wasm byte size
   based on that. For very small wasm functions, large threshold will
   reduce use of threads of the system. But for enough sized wasm
   binary, we should bulk compile many functions to reduce rescheduling
   of wasm plan cost.
3. We also remove unnecessary locking in WasmIPIntPlan / WasmLLIntPlan.

                                    ToT                     Patched

    richards-wasm-load       569.6911+-26.6883    ^    385.5147+-5.0021        ^ definitely 1.4777x faster
    quicksort-wasm-load      508.5815+-0.6967     ^    381.4527+-3.4712        ^ definitely 1.3333x faster
    gcc-loops-wasm-load      263.1294+-5.8862     ?    264.7556+-2.9799        ?
    hashset-wasm-load        402.5893+-2.0566     ^    381.7596+-0.8357        ^ definitely 1.0546x faster
    tsf-wasm-load            351.2520+-5.2388     ^    303.8165+-1.9450        ^ definitely 1.1561x faster

* JSTests/slowMicrobenchmarks/gcc-loops-wasm-load.js: Added.
* JSTests/slowMicrobenchmarks/quicksort-wasm-load.js: Added.
* JSTests/slowMicrobenchmarks/richards-wasm-load.js: Added.
* JSTests/slowMicrobenchmarks/tsf-wasm-load.js: Added.
* Source/JavaScriptCore/runtime/OptionsList.h:
* Source/JavaScriptCore/wasm/WasmEntryPlan.cpp:
(JSC::Wasm::EntryPlan::compileFunctions):
* Source/JavaScriptCore/wasm/WasmFunctionCodeBlockGenerator.cpp:
(JSC::Wasm::FunctionCodeBlockGenerator::setTailCall):
* Source/JavaScriptCore/wasm/WasmFunctionCodeBlockGenerator.h:
(JSC::Wasm::FunctionCodeBlockGenerator::hasTailCallSuccessors const):
* Source/JavaScriptCore/wasm/WasmFunctionIPIntMetadataGenerator.cpp:
(JSC::Wasm::FunctionIPIntMetadataGenerator::setTailCall):
* Source/JavaScriptCore/wasm/WasmFunctionIPIntMetadataGenerator.h:
(JSC::Wasm::FunctionIPIntMetadataGenerator::hasTailCallSuccessors const):
* Source/JavaScriptCore/wasm/WasmIPIntPlan.cpp:
(JSC::Wasm::IPIntPlan::compileFunction):
* Source/JavaScriptCore/wasm/WasmIPIntPlan.h:
* Source/JavaScriptCore/wasm/WasmLLIntPlan.cpp:
(JSC::Wasm::LLIntPlan::compileFunction):
* Source/JavaScriptCore/wasm/WasmLLIntPlan.h:
* Source/JavaScriptCore/wasm/WasmModuleInformation.h:
* Source/JavaScriptCore/wasm/WasmStreamingParser.cpp:
(JSC::Wasm::StreamingParser::parseFunctionPayload):
* Source/JavaScriptCore/wasm/WasmStreamingParser.h:

Canonical link: <a href="https://commits.webkit.org/289816@main">https://commits.webkit.org/289816@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/1049026cc2afd99122302a85661126c6b0d0a497

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/88081 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/7597 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/42482 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/93029 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/38831 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/7978 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/15775 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/67990 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/25721 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/91083 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/6110 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/79686 "Passed tests") | [❌ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/48353 "Found 1 new API test failure: /TestJSC:/jsc/options (failure)") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/5884 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/34100 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/37939 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/80879 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/76273 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/34978 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/94877 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/86857 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/15249 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/11205 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/76851 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/15504 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/75542 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/76091 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/20475 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/119/builds/18882 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/8278 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/13747 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/15267 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/20568 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/109350 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/15009 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/26296 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/18454 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/16791 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

kmiller68 commented on Source/JavaScriptCore/wasm/WasmEntryPlan.cpp:
```diff
@@ -200,11 +198,26 @@ void EntryPlan::compileFunctions(CompilationEffort effort)
+                }
+                bytesCompiled += byteSize;
+                if (bytesCompiled >= compileLimit) {
                     functionIndexEnd = index + 1;
```
Why do any of this calculation if `effort == All`? We're going to loop around and ignore it anyway.

Constellation replied:
In fact, it turned out that nobody is using `effort == All` anymore :) I'll remove this `effort` parameter. Will remove them.

kmiller68 reviewed (approved):
r=me

hyjorc1 reviewed (approved):
r=me

Constellation committed:
[JSC] More sophisticated wasm worklist distribution
https://bugs.webkit.org/show_bug.cgi?id=286986
rdar://144136820

Reviewed by Keith Miller and Yijia Huang.

This patch tweaks our wasm worklist distribution code for non streaming
wasm compilation.

1. We check each function size and isolate super huge function in a
   different loop of wasm compilation so that we can distribute outlier
   sized function separately from the rest, which makes distribution of
   compilation task more uniformly.
2. We compute total size of functions and change partial wasm byte size
   based on that. For very small wasm functions, large threshold will
   reduce use of threads of the system. But for enough sized wasm
   binary, we should bulk compile many functions to reduce rescheduling
   of wasm plan cost.
3. We also remove unnecessary locking in WasmIPIntPlan / WasmLLIntPlan.

                                    ToT                     Patched

    richards-wasm-load       569.6911+-26.6883    ^    385.5147+-5.0021        ^ definitely 1.4777x faster
    quicksort-wasm-load      508.5815+-0.6967     ^    381.4527+-3.4712        ^ definitely 1.3333x faster
    gcc-loops-wasm-load      263.1294+-5.8862     ?    264.7556+-2.9799        ?
    hashset-wasm-load        402.5893+-2.0566     ^    381.7596+-0.8357        ^ definitely 1.0546x faster
    tsf-wasm-load            351.2520+-5.2388     ^    303.8165+-1.9450        ^ definitely 1.1561x faster

* JSTests/slowMicrobenchmarks/gcc-loops-wasm-load.js: Added.
* JSTests/slowMicrobenchmarks/quicksort-wasm-load.js: Added.
* JSTests/slowMicrobenchmarks/richards-wasm-load.js: Added.
* JSTests/slowMicrobenchmarks/tsf-wasm-load.js: Added.
* Source/JavaScriptCore/runtime/OptionsList.h:
* Source/JavaScriptCore/wasm/WasmEntryPlan.cpp:
(JSC::Wasm::EntryPlan::compileFunctions):
* Source/JavaScriptCore/wasm/WasmFunctionCodeBlockGenerator.cpp:
(JSC::Wasm::FunctionCodeBlockGenerator::setTailCall):
* Source/JavaScriptCore/wasm/WasmFunctionCodeBlockGenerator.h:
(JSC::Wasm::FunctionCodeBlockGenerator::hasTailCallSuccessors const):
* Source/JavaScriptCore/wasm/WasmFunctionIPIntMetadataGenerator.cpp:
(JSC::Wasm::FunctionIPIntMetadataGenerator::setTailCall):
* Source/JavaScriptCore/wasm/WasmFunctionIPIntMetadataGenerator.h:
(JSC::Wasm::FunctionIPIntMetadataGenerator::hasTailCallSuccessors const):
* Source/JavaScriptCore/wasm/WasmIPIntPlan.cpp:
(JSC::Wasm::IPIntPlan::compileFunction):
* Source/JavaScriptCore/wasm/WasmIPIntPlan.h:
* Source/JavaScriptCore/wasm/WasmLLIntPlan.cpp:
(JSC::Wasm::LLIntPlan::compileFunction):
* Source/JavaScriptCore/wasm/WasmLLIntPlan.h:
* Source/JavaScriptCore/wasm/WasmModuleInformation.h:
* Source/JavaScriptCore/wasm/WasmStreamingParser.cpp:
(JSC::Wasm::StreamingParser::parseFunctionPayload):
* Source/JavaScriptCore/wasm/WasmStreamingParser.h:

Canonical link: https://commits.webkit.org/289816@main
